kernel/src/hil/usb_hid.rs

```diff
@@ -0,0 +1,86 @@
+//! Interface for USB HID (Human Interface Device) class
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// Called when a packet is received.
+    /// This will return the buffer passed into `receive_buffer()` as well as
+    /// the endpoint where the data was received. The `len` argument contains
+    /// the length of the packet received. If the buffer length is smaller
+    /// then this length the buffer will only contain part of the packet and
+    /// `result` will contain indicate an `ESIZE` error.
+    fn packet_received(
+        &'a self,
+        result: ReturnCode,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when a packet has been finished transmitting.
+    /// This will return the buffer passed into `send_buffer()` as well as
+    /// the endpoint where the data was sent. The `len` argument contains
+    /// the number of bytes sent. If this is less then the size of the
+    /// buffer `result` will contain the `ESIZE` error.
+    fn packet_transmitted(
+        &'a self,
+        result: ReturnCode,
+        buffer: &'static mut [u8; 64],
+        len: usize,
+        endpoint: usize,
+    );
+
+    /// Called when checking if we can start a new receive operation.
+    /// Should return true if we are ready to receive and not currently
+    /// in the process of receiving anything. That is if we are currently
+    /// idle.
+    /// If there is an outstanding call to receive, a callback already
+    /// waiting to be called then this will return false.
+    fn can_receive(&'a self) -> bool;
```

Comment:
Given the synchronous nature of the Tock kernel, I don't think you can have TOCTOU bugs; it would have to be you receive a callback preemptively. 

The presence of a call like this allows the client to push out replicated internal state. Suppose the client wants to know if there's currently a reception (e.g., whether it can power something down). It can replicate this part of the USB state machine internally by tracking calls to receive, or it can rely on calling this method. It might want to know the state without starting a new reception.